Feature/filtering support

mattdelsordo opened the pull request:
## Overview

This PR adds layer filtering functionality. To quote what I put in the README:

> Map data can be filtered by specifying a query in the form `( QUERY ) as VARIABLE` as the value of `table` for a layer in `map-config.mml`. Different combinations of layers can be fetched by adding `?layers=layer1,layer2,layer...N` as a query string to a tile or UTF grid URL. Make sure that all necessary columns are present in the query. For example, if your geometry column is called `geom`, and you want to display a UTF grid of content from the column `name` filtered by the content of the column `type`, your query would look something like `SELECT geom,name,type FROM table WHERE type='something'`. If no layers are specified in the query string, all layers are displayed in the same tile.



### Demo

![image](https://user-images.githubusercontent.com/16990679/42388778-a7bfa992-8114-11e8-841a-7ca778fb6813.png)
![image](https://user-images.githubusercontent.com/16990679/42388815-baafc7d0-8114-11e8-9f56-25299881aea1.png)

### Notes

There's a persisting shearing issue in tiles with large numbers of markers, this will be addressed in a subsequent PR.


## Testing Instructions

 * Run the development server as per usual.
 * Open up the demo webpage in `/src/demo`. It already has several example filters enabled on it.
 * To add a filer to the demo, you have to alter the following:
   * `src/tiler/src/config/map-config.mml`: Add a new layer with a unique name and a new query as the value of `table`. For your convenience, here's the values of `owner` that I left out of the demo: 
     * PRIVPWDMAINT
     * CHEL
     * PRIVMAND
     * STREET
     * BENS
     * HVF
     * PIDC
     * LML
     * SPR
     * AUTH
     * ABT
    * `src/tiler/src/config/style.mss`: Add a new style class with the same name as the layer you made and style it however you want.
    * `src/demo/pwd_inlets/index.html`: In the `#layer-selector` span, add a new input/label cluster for your layer. Then, in the script section, add a new `createLayer('$NAME_OF_YOUR_LAYER')`.
  * Reload the demo page and check your layer's box to see if it is properly served up.

Resolves #27

mattdelsordo committed:
Add layer filtering functionality

mattdelsordo committed:
Improve filtering documentation

mattdelsordo committed:
Enable filtering UTF grids by layer

mattdelsordo committed:
Fix filtering happening the reverse of how intended

mattdelsordo committed:
Update demo to show off filtering and tiling

mattdelsordo committed:
Update readme to inform about filtering behavior

KlaasH commented on node_lambnik/README.md:
```diff
@@ -6,15 +6,18 @@
 [Tilegarden uses CartoCSS to specify map styles.](https://carto.com/docs/carto-engine/cartocss/) Edit `src/tiler/src/config/map-config.mml` to specify your map configuration, including a reference any CartoCSS `.mss` files. [See here](https://cartocss.readthedocs.io/en/latest/mml.html) for Carto's `.mml` specification.
 
+### Filtering
+Map data can be filtered by specifying a query in the form `( QUERY ) as VARIABLE` as the value of `table` for a layer in `map-config.mml`. Different combinations of layers can be fetched by adding `?layers=layer1,layer2,layer...N` as a query string to a tile or UTF grid URL. Make sure that all necessary columns are present in the query, for example, if your geometry column is called `geom`, and  ...
```
Copy notes: I would start a new sentence at "For example", and there's an extra quote on the example query.

KlaasH commented on node_lambnik/src/demo/pwd_inlets_index.html:
```diff
@@ -35,39 +62,71 @@ <h3>PWD Inlets</h3>
+        createLayer('AIRPRT');
+        createLayer('LM');
+        createLayer('FEDERAL');
+        createLayer('PRIV');
```
It's just code on a demo page, but the repetition here and the indirectness above are bothering me a little.  The latter especially, since I would expect `createLayer` to return the layer rather than modify an object from outside its scope.

I would vote for changing the function to return the layer and changing the object creation to:
```javascript
var layerKeys = ['PWD', 'STATE', 'PARK', 'AIRPRT', 'LM', 'FEDERAL', 'PRIV']
const layers = Object.assign(...layerKeys.map(key => ({[key]: createLayer(key)})));
```
(and using `layers` below rathar than `layerKeys`. more accurately descriptive)
That's an annoying number of closing parens with a curly brace thrown in for extra unreadability, but it has the advantage that it makes the object directly rather than as a side-effect.  And it's short.

KlaasH commented on node_lambnik/src/tiler/src/tiler.js:
```diff
@@ -17,21 +17,42 @@ const TILE_WIDTH = 256
+                if (!layers.includes(layerName)) {
+                    return `${beginning} status="0">`
+                }
+                return fullMatch
```
I think trying to parse markup with regexes is always a dangerous game, and the fact that this very simple attribute addition is not particularly straightforward to read and already seems kind of brittle (can `Layer` have other attributes besides `name`?  If so, this would stop working.) is a decent illustration of why.  I think we should find a (hopefully pretty lightweight, since our needs are quite modest) package to load this as XML and let us traverse it and set attributes as such, then write it back out to a string.

mattdelsordo replied:
I agree, I've realized when writing tests that it already doesn't work even in the simple use cases we've been demoing it on. I'll go swap this out for an XML library, we might need one for YAML too.

KlaasH commented on node_lambnik/src/tiler/src/tiler.js:
```diff
@@ -17,21 +17,42 @@ const TILE_WIDTH = 256
 
     // Load map specification from xml string
     return readFile(path.join(__dirname, 'map-config.xml'), 'utf-8')
+        .then(xml => filterLayers(xml, layers))
```
Is it really true that layers can't be dropped after the map object is made?  It certainly [looks that way](http://mapnik.org/documentation/node-mapnik/3.6/#Map), but it seems like such obviously useful functionality...

mattdelsordo replied:
Yeah, I agree, it seems like a weird design choice to me. [I found this issue in node-mapnik](https://github.com/mapnik/node-mapnik/issues/742) where someone expresses a need for the same functionality, and Dane Springmeyer (who's the chief contributor to the project) said the following:
> This is a known limitation that I don't think should be changed. I don't think an API that allows mutating maps in place is within scope. I recommend thinking about map instances as best created once and then rendered one or more times and never modified again.

mattdelsordo committed:
Add XML filtering module that uses a real XML parser

https://stackoverflow.com/a/1732454

mattdelsordo committed:
Make yml templating function slightly more robust

It is unclear at this time whether this method warrants replacement
with a full-on YAML parser, as the pattern that's getting searched for
isn't tied to YAML syntax itself.

KlaasH reviewed (approved):
Looks great! :+1:
